go/tools/releaser/upgradedep.go

```diff
@@ -0,0 +1,518 @@
+// Copyright 2021 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	bzl "github.com/bazelbuild/buildtools/build"
+	"github.com/google/go-github/v36/github"
+	"golang.org/x/mod/semver"
+	"golang.org/x/oauth2"
+	"golang.org/x/sync/errgroup"
+)
+
+var upgradeDep = command{
+	name:        "upgrade-dep",
+	run:         runUpgradeDep,
+	description: "upgrades a dependency in WORKSPACE or go_repositories.bzl",
+	help: `releaser upgrade-dep [-githubtoken=token] [-work] deps...
+
+upgrade-dep upgrades one or more rules_go dependencies in WORKSPACE or
+go/private/repositories.bzl. Dependency names (matching the name attributes)
+can be specified with positional arguments. "all" may be specified to upgrade
+all upgradeable dependencies.
+
+For each dependency, upgrade-dep finds the highest version available in the
+upstream repository. If no version is available, upgrade-dep uses the commit
+at the tip of the default branch. If a version is part of a release,
+upgrade-dep will try to use an archive attached to the release; if none is
+available, upgrade-dep uses an archive generated by GitHub.
+
+Once upgrade-dep has found the URL for the latest version, it will:
+
+* Download the archive.
+* Upload the archive to mirror.bazel.build.
+* Re-generate patches, either by running a command or by re-applying the
+  old patches.
+* Update dependency attributes in WORKSPACE and repositories.bzl, then format
+  and rewrite those files.
+
+Upgradeable dependencies need a comment like '# releaser:upgrade-dep org repo'
```

Comment:
Should this work the other way? Is it possible to default to things being upgradable unless otherwise specified?

Reply:
It's probably a good idea eventually, but I'd like to gain more confidence in this approach first. Supporting `go_repository` and `git_repository` first would be good.

This is just meant to support rules_go and Gazelle maintenance so I'm not too worried about changing the directives at this point.